dev/src/bulk-writer.ts

```diff
@@ -714,27 +748,34 @@ export class BulkWriter {
     const tag = requestTag();
     const pendingBatch = this._bulkCommitBatch;
 
-    this._bulkCommitBatch = new BulkCommitBatch(this.firestore);
-
-    // Send the batch if it is under the rate limit, or schedule another
-    // attempt after the appropriate timeout.
-    const underRateLimit = this._rateLimiter.tryMakeRequest(
+    // Use the write with the highest failure count when determining backoff.
+    const highestBackoffDuration = pendingBatch.pendingOps.reduce((prev, cur) =>
+      prev.backoffDuration > cur.backoffDuration ? prev : cur
+    ).backoffDuration;
+    const backoffMsWithJitter = BulkWriter.applyJitter(highestBackoffDuration);
+    const delayMs = this._rateLimiter.getNextRequestDelayMs(
       pendingBatch._opCount
     );
 
+    this._bulkCommitBatch = new BulkCommitBatch(this.firestore);
+
     const delayedExecution = new Deferred<void>();
-    if (underRateLimit) {
+    // Send the batch if it is does not require any delay, or schedule another
+    // attempt after the appropriate timeout.
+    if (backoffMsWithJitter === 0 && delayMs === 0) {
+      assert(
+        this._rateLimiter.tryMakeRequest(pendingBatch._opCount),
+        'RateLimiter should allow request if delayMs === 0'
+      );
       delayedExecution.resolve();
     } else {
-      const delayMs = this._rateLimiter.getNextRequestDelayMs(
-        pendingBatch._opCount
-      );
+      const finalDelayMs = Math.max(backoffMsWithJitter, delayMs);
```

Comment:
You can move out of the if statement and then use it in the `if(...)` check itself.

Reply:
Done.